flink-core/src/main/java/org/apache/flink/core/io/SimpleVersionedSerializerTypeSerializerProxy.java

```diff
@@ -0,0 +1,163 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.io;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.function.Supplier;
+
+/**
+ * A {@link TypeSerializer} that delegates to an underlying {@link SimpleVersionedSerializer}.
+ */
+@Internal
+public class SimpleVersionedSerializerTypeSerializerProxy<T> extends TypeSerializer<T> implements Serializable {
+
+	private final SerializableSupplier<SimpleVersionedSerializer<T>> serializerSupplier;
+	private transient SimpleVersionedSerializer<T> cachedSerializer;
+
+	public SimpleVersionedSerializerTypeSerializerProxy(
+			SerializableSupplier<SimpleVersionedSerializer<T>> serializerSupplier) {
+		this.serializerSupplier = serializerSupplier;
+	}
+
+	@Override
+	public boolean isImmutableType() {
+		return false;
+	}
+
+	@Override
+	public TypeSerializer<T> duplicate() {
+		final byte[] serializedSerializer;
+		try {
+			serializedSerializer = InstantiationUtil.serializeObject(serializerSupplier);
+		} catch (IOException e) {
+			throw new RuntimeException("Could not serialize SimpleVersionedSerializer.", e);
+		}
+		try {
+			return new SimpleVersionedSerializerTypeSerializerProxy<>(
+					InstantiationUtil.deserializeObject(
+							serializedSerializer,
+							serializerSupplier.getClass().getClassLoader()));
+		} catch (ClassNotFoundException | IOException e) {
+			throw new RuntimeException("Could not duplicate SimpleVersionedSerializer.", e);
+		}
+	}
+
+	@Override
+	public T createInstance() {
+		return null;
+	}
+
+	@Override
+	public T copy(T from) {
+		SimpleVersionedSerializer<T> serializer = getSerializer();
+		try {
+			byte[] serializedFrom = serializer.serialize(from);
+			return serializer.deserialize(
+					serializer.getVersion(),
+					serializedFrom);
+		} catch (IOException e) {
+			throw new RuntimeException("Could not copy element.", e);
+		}
+	}
+
+	@Override
+	public T copy(T from, T reuse) {
+		// the reuse is optional, we can just ignore it
+		return copy(from);
+	}
+
+	@Override
+	public int getLength() {
+		return -1;
+	}
+
+	@Override
+	public void serialize(T record, DataOutputView target) throws IOException {
+		SimpleVersionedSerializer<T> serializer = getSerializer();
+		target.writeInt(serializer.getVersion());
```

Comment:
Do we need to write the version of the serializer? If I am not mistaken it will be used only for the network transfer and thus always work with a single version of the serializer.

Reply:
You are right. I tried to be as general as possible just in case someone came upon this code and also tried to use it. Also, with unaligned checkpoints in-flight records are checkpointed so it becomes relevant for that. Although we currently say that UC cannot be used for version upgrades (of both Flink or the user job).

Reply:
Hah, I actually removed a part about unaligned checkpoints from my initial comment ;) I dropped it when I actually realized that we don't support unaligned savepoints and thus version upgrades. (or am I wrong?)

Shouldn't we then implement the `snapshotConfiguration`? This method is mandatory if we want to consider version upgrades, imo...

Reply:
Yes, this whole things a pickle. I don't want to add a snapshot because that's complicated and would require me to serialize the supplier, which we shouldn't do.

I'm fine with removing the version but I think leaving it in would make the code a tiny bit more future proof.

Reply:
@aljoscha I have concerned whether `TypeSerializerSnapshot` need `SimpleVersionedTypeSerializerSnapshotProxy` like this. What do you think about this?

Reply:
Right now the serializer will be just for `Committables` in the Sink interface so having the version won't harm as there won't be much data volume. If it is used for regular additional version might be significant.

Having said that I am fine with either, leaving or removing.

Reply:
@SteNicholas What do you mean by that?

Reply:
> @SteNicholas What do you mean by that?

I just want to ask whether `TypeSerializerSnapshot ` should be handled similarly with `TypeSerializer`.

Reply:
No, I don't want to do that right now because this is not meant as a general `TypeSerializer` that users should use for state. I'm expressly forbidding that by throwing an exception.